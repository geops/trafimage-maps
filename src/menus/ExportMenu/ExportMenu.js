import React, { useState, useRef, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { FaDownload } from 'react-icons/fa';
import { makeStyles } from '@material-ui/core';
import MenuItem from '../../components/Menu/MenuItem';
import ExportButton from '../../components/ExportButton';
import determineMaxCanvasSize from '../../utils/canvasSize';

const LS_SIZE_KEY = 'tm.max.canvas.size';

const useStyles = makeStyles(() => ({
  menuContent: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
  },
}));

const ExportMenu = () => {
  const classes = useStyles();
  const [collapsed, setCollapsed] = useState(false);
  const [maxCanvasSize, setMaxCanvasSize] = useState(
    localStorage.getItem(LS_SIZE_KEY),
  );
  const permissionInfos = useSelector((state) => state.app.permissionInfos);
  const { t } = useTranslation();
  const ref = useRef();

  useEffect(() => {
    let timeout = null;
    if (maxCanvasSize) {
      return () => {};
    }
    timeout = setTimeout(() => {
      const size = determineMaxCanvasSize();
      if (size) {
        setMaxCanvasSize(size);
        localStorage.setItem(LS_SIZE_KEY, size);
      }
    }, 10);
    return () => {
      clearTimeout(timeout);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (!permissionInfos || !permissionInfos.user || !maxCanvasSize) {
    return null;
  }

  return (
    <div ref={ref} id="test">
      <MenuItem
        open
        title={t('Grossformatiges PDF exportieren')}
        icon={<FaDownload focusable={false} />}
        collapsed={collapsed}
        onCollapseToggle={(c) => setCollapsed(c)}
      >
        <div className={classes.menuContent}>
          <ExportButton maxCanvasSize={maxCanvasSize}>A0 (72dpi)</ExportButton>
          <ExportButton exportScale={2} maxCanvasSize={maxCanvasSize}>
            A0 (150dpi)
          </ExportButton>
          <ExportButton exportScale={3} maxCanvasSize={maxCanvasSize}>
            A0 (300dpi)
          </ExportButton>
          <ExportButton exportFormat="a1" maxCanvasSize={maxCanvasSize}>
            A1 (72dpi)
          </ExportButton>
          <ExportButton
            exportFormat="a1"
            exportScale={2}
            maxCanvasSize={maxCanvasSize}
          >
            A1 (150dpi)
          </ExportButton>
          <ExportButton
            exportFormat="a1"
            exportScale={3}
            maxCanvasSize={maxCanvasSize}
          >
            A1 (300dpi)
          </ExportButton>
        </div>
      </MenuItem>
    </div>
  );
};

export default React.memo(ExportMenu);
