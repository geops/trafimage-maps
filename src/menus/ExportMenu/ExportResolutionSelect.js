import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { makeStyles } from "@mui/styles";
import { MenuItem, InputLabel, FormControl } from "@mui/material";
import { useTranslation } from "react-i18next";
import { useDispatch, useSelector } from "react-redux";
import Select from "../../components/Select";
import { validateOption } from "../../utils/exportUtils";
import { setExportPrintOptions } from "../../model/app/actions";
import useMaxCanvasSize from "../../utils/useMaxCanvasSize";
import useExportPrintOptions from "../../utils/useExportPrintOptions";

const useStyles = makeStyles(() => ({
  label: {
    top: -8,
    left: -12,
  },
}));

function ExportResolutionSelect({ options, className }) {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const classes = useStyles();
  const map = useSelector((state) => state.app.map);
  const maxCanvasSize = useMaxCanvasSize();
  const exportPrintOptions = useExportPrintOptions(options);
  const [value, setValue] = useState(exportPrintOptions);

  useEffect(() => {
    if (exportPrintOptions) {
      setValue(
        options.find(
          (opt) =>
            opt.quality === exportPrintOptions.quality &&
            opt.paperSize === exportPrintOptions.paperSize,
        ),
      );
    }
  }, [exportPrintOptions, options]);

  if (!exportPrintOptions || !value) return null;

  return (
    <FormControl fullWidth className={className}>
      <InputLabel className={classes.label}>{t("Format")}</InputLabel>
      <Select
        id="pdf-format-select-label"
        className={classes.input}
        value={value}
        onChange={(evt) => dispatch(setExportPrintOptions(evt.target.value))}
      >
        {options.map((opt) => {
          return (
            <MenuItem
              key={`${opt.paperSize}-${opt.quality}`}
              value={opt}
              disabled={validateOption(
                `${opt.paperSize}`,
                opt.quality,
                maxCanvasSize,
                map,
              )}
            >
              {opt.label}
            </MenuItem>
          );
        })}
      </Select>
    </FormControl>
  );
}

ExportResolutionSelect.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      paperSize: PropTypes.string.isRequired,
      quality: PropTypes.number.isRequired,
    }),
  ).isRequired,
  className: PropTypes.string,
};

ExportResolutionSelect.defaultProps = {
  className: "",
};

export default ExportResolutionSelect;
