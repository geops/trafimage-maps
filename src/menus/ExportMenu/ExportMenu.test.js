import 'jest-canvas-mock';
import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { Map, View } from 'ol';
import ExportMenu from '.';

configure({ adapter: new Adapter() });

describe('ExportMenu', () => {
  const mockStore = configureStore([thunk]);

  test('should match snapshot when not logged in.', () => {
    const store = mockStore({
      app: {
        menuOpen: true,
        map: new Map({ view: new View({}) }),
        permissionInfos: null,
      },
    });

    const component = renderer.create(
      <Provider store={store}>
        <ExportMenu />
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match snapshot when logged in.', () => {
    const store = mockStore({
      app: {
        menuOpen: true,
        map: new Map({ view: new View({}) }),
        permissionInfos: {
          user: {},
        },
      },
    });

    const component = renderer.create(
      <Provider store={store}>
        <ExportMenu />
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
