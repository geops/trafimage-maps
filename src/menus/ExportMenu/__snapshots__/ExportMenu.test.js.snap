// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportMenu should match snapshot when logged in. 1`] = `null`;

exports[`ExportMenu should match snapshot when not logged in. 1`] = `null`;
